### PR TITLE
fix: handle new xchain gauges, fix harvests

### PR DESCRIPTION
**Pull request by btcookies**
Curve has added support for cross chain boosted gauges, changing the architecture in the process. This is breaking harvests because the gauge itself no longer contains CRV. Based on architecture explained [here](https://github.com/curvefi/curve-xchain-factory) in order to claim a LP's earned CRV the address must call `GaugeFactory.mint(gaugeAddress)`.

Updates required to harvest logic to support this. This PR updates:
- new storage variable `gaugeFactory` that can be changed
- new setter `setGaugeFactory(address)` to support updating `gaugeFactory` permissioned to `_onlyGovernance()`
- `harvest()` updated to call `ICurveGaugeFactory(gaugeFactory).mint(gauge);` to realize earned CRV

### CODE REVIEW

**Comment by btcookies** [comment]
Screen Shot 2022-05-18 at 12 27 41 PM

File: tests/test_upgrade_gauge_factory.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    gauge = interface.ICurveGauge(NEW_GAUGE)
+    
+    want = interface.IERC20(WANT)
+    prev_bal = gauge.balanceOf(strat_proxy.address)
+
+    brownie.chain.sleep(60*60*24)
+    brownie.chain.mine()
+
+    ## Now it will work
+    strat_proxy.harvest({"from": gov})
```
Comment (sajanrajdev) [inline tests/test_upgrade_gauge_factory.py]: This doesn't really ensure that Harvests work properly. You can use the Snapshot manager here and call harvest as done in this harvest flow test: https://github.com/GalloDaSballo/arbitrum-ren-wbtc-curve-strat/blob/af8057b95f8015160dd708dc2d4de479d8cf0531/tests/examples/test_harvest_flow.py#L87

By calling harvest from the SnapshotManager, it atomatically runs all of the checks from the expected behavior of the Strat's harvest function.

File: contracts/MyStrategy.sol [hunk 2]
```diff
@@ -205,7 +216,7 @@ contract MyStrategy is BaseStrategy {
         uint256 _before = IERC20Upgradeable(want).balanceOf(address(this));
 
         // figure out and claim our rewards
-        ICurveGauge(gauge).claim_rewards();
+        ICurveGaugeFactory(gaugeFactory).mint(gauge);
```
Comment (sajanrajdev) [inline contracts/MyStrategy.sol:219]: The GaugeFactory from the test doesn't contain a .mint() function: https://etherscan.io/address/0xabC000d88f23Bb45525E447528DBF656A9D55bf5#writeContract

image

File: tests/test_upgrade_gauge_factory.py (hunk 1 above)
Comment (sajanrajdev) [reply tests/test_upgrade_gauge_factory.py:101]: Isn't it supposed to be `transmit_emissions`?

File: contracts/MyStrategy.sol (hunk 2 above)
Comment (sajanrajdev) [reply contracts/MyStrategy.sol:219]: Isn't it supposed to be `transmit_emissions`?

Comment (sajanrajdev) [reply contracts/MyStrategy.sol:219]: It seems like the address on the test points to the rootGaugeFactory and we need the childGaugeFactory

Comment (sajanrajdev) [reply contracts/MyStrategy.sol:219]: Nevermind, sorry, was looking at Etherscan... The same address on Arbiscan is the ChildGaugeFactory indeed: https://arbiscan.io/address/0xabC000d88f23Bb45525E447528DBF656A9D55bf5, very weird...

Comment (sajanrajdev) [reply contracts/MyStrategy.sol:219]: Confirm that `factory.mint(gauge)` indeed mints the amount accumulated by the user (strat) on the gauge in matter and transfers the amount directly to the user: https://github.com/curvefi/curve-xchain-factory/blob/6066c873713c50c6baaacad460b7be48fbd45846/contracts/ChildGaugeFactory.vy#L107-L113

Not very intuitive. Very good finding!

File: tests/test_upgrade_gauge_factory.py (hunk 1 above)
Comment (sajanrajdev) [reply tests/test_upgrade_gauge_factory.py]: In other words, it is important to test not only that operational functions run but also that they run properly. You can add manual checks for the ppfs and strat balance increasing after harvest o you can use the SnapshotManager to have this checked automatically.

Comment (sajanrajdev) [reply tests/test_upgrade_gauge_factory.py]: See the resolver to explore the checks that the manager executes: https://github.com/GalloDaSballo/arbitrum-ren-wbtc-curve-strat/blob/af8057b95f8015160dd708dc2d4de479d8cf0531/config/StrategyResolver.py#L44-L70

File: contracts/MyStrategy.sol (hunk 2 above)
Comment (btcookies) [reply contracts/MyStrategy.sol:219]: arbiscan

**Review by sajanrajdev — COMMENTED** [review]
Harvest simulation post upgrade shows rewards properly being received and handled:
image

@btcookies, some of the other tests seem to be broken, they revert on harvest. I think you need to call `strat.setFactory()` within the confest setup so that the contract is loaded on the rest of the tests and harvests work there:
image

**Review by sajanrajdev — APPROVED** [review]
All tests are passing now:
image
Re-ran failed test: 
image

Changes LGTM

**Comment by GalloDaSballo** [comment]
Code LGTM

Upgrade code works and rationally checks for storage conflicts.

Tests are passing exactly as shown by Saj (failing automatic test)
Screenshot 2022-05-19 at 01 54 07

Running test separately makes it pass
Screenshot 2022-05-19 at 02 00 34

**Pull request merged**